core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java

```diff
@@ -498,76 +584,114 @@ long calculateTimeoutSeconds() {
-         this.keySet = keySet;
+      public WrappedKeySet(KeySetCommand command, long remoteIterationFlags, CacheSet<K> keySet) {
+         super(keySet);
+         this.cache = Caches.getCacheWithFlags(CacheLoaderInterceptor.this.cache, command);
+         this.isRemoteIteration = remoteIterationFlags == FlagBitSets.REMOTE_ITERATION;
+      }
+
+      @Override
+      public boolean remove(Object o) {
+         // Remove must be done by the cache
+         return o != null && cache.remove(o) != null;
       }
 
       @Override
       public CloseableIterator<K> iterator() {
-         CloseableIterator<K> iterator = Closeables.iterator(keySet.stream());
+         if (isRemoteIteration) {
+            return innerIterator();
+         }
+         // Need to support remove of iterator
+         return new RemovableCloseableIterator<>(innerIterator(), cache::remove);
+      }
+
+      @Override
+      protected CloseableIterator<K> innerIterator() {
+         CloseableIterator<K> iterator = cacheSet.iterator();
          Set<K> seenKeys = new HashSet<>(cache.getAdvancedCache().getDataContainer().sizeIncludingExpired());
          // TODO: how to handle concurrent activation....
-         return new DistinctKeyDoubleEntryCloseableIterator<>(iterator, new CloseableSuppliedIterator<>(new SupplierFunction<>(
-               new PersistenceManagerCloseableSupplier<>(executorService, persistenceManager,
-                     // TODO: how to pass in key filter...
-                     iceFactory, new CollectionKeyFilter<>(seenKeys), 10, TimeUnit.SECONDS, 2048, false))),
+         return new DistinctKeyDoubleEntryCloseableIterator<>(iterator, new CloseableSuppliedIterator<>(
+               new SupplierFunction<>(// TODO: how to pass in key filter...
+                     new PersistenceManagerCloseableSupplier<>(executorService, persistenceManager, iceFactory,
+                           new CollectionKeyFilter<>(seenKeys), 10, TimeUnit.SECONDS, 2048, false))),
                Function.identity(), seenKeys);
       }
 
       @Override
-      public CloseableSpliterator<K> spliterator() {
-         return spliteratorFromIterator(iterator());
-      }
-
-      private <E> CloseableSpliterator<E> spliteratorFromIterator(CloseableIterator<E> iterator) {
-         return new IteratorAsSpliterator.Builder<>(iterator).setCharacteristics(
-               Spliterator.CONCURRENT | Spliterator.DISTINCT | Spliterator.NONNULL).get();
+      protected AbstractLocalCacheStream.StreamSupplier<K, Stream<K>> supplier() {
+         return new KeyStreamSupplier<>(cache, getSegmentMapper(cache), () -> StreamSupport.stream(spliterator(), false));
       }
 
       @Override
-      public int size() {
-         long size = stream().count();
-         if (size > Integer.MAX_VALUE) {
-            return Integer.MAX_VALUE;
+      public boolean contains(Object o) {
+         boolean contains = super.contains(o);
```

Comment:
if o is null, should no we just return false ?

Reply:
Sure, in the Collection API it is valid, but we don't allow null keys.